cpp/include/cudf/detail/utilities/dremel.hpp

```diff
@@ -0,0 +1,195 @@
+ * is an encoding that converts record-like data into a columnar store for efficient memory
+ * accesses. The Parquet file format uses Dremel encoding to handle nested data, so libcudf
+ * requires some facilities for working with this encoding. Furthermore, libcudf leverages
+ * Dremel encoding as a means for performing lexicographic comparisons of nested columns.
+ *
+ * Dremel encoding is built around two concepts, the repetition and definition levels.
+ * Since describing them thoroughly is out of scope for this docstring, here are a couple of
+ * blogs that provide useful background:
+ * http://www.goldsborough.me/distributed-systems/2019/05/18/21-09-00-a_look_at_dremel/
+ * https://akshays-blog.medium.com/wrapping-head-around-repetition-and-definition-levels-in-dremel-powering-bigquery-c1a33c9695da
+ *
+ * The remainder of this documentation assumes familiarity with the Dremel concepts.
+ *
+ * Dremel offsets are the per row offsets into the repetition and definition level arrays for a
+ * column.
+ * Example:
+ * ```
+ * col            = {{1, 2, 3}, { }, {5, 6}}
+ * dremel_offsets = { 0,         3,   4,  6}
+ * rep_level      = { 0, 1, 1,   0,   0, 1}
+ * def_level      = { 1, 1, 1,   0,   1, 1}
+ * ```
+ *
+ * The repetition and definition level values are ideally computed using a recursive call over a
+ * nested structure but in order to better utilize GPU resources, this function calculates them
+ * with a bottom up merge method.
+ *
+ * Given a LIST column of type `List<List<int>>` like so:
+ * ```
+ * col = {
+ *    [],
+ *    [[], [1, 2, 3], [4, 5]],
+ *    [[]]
+ * }
+ * ```
+ * We can represent it in cudf format with two level of offsets like this:
+ * ```
+ * Level 0 offsets = {0, 0, 3, 5, 6}
+ * Level 1 offsets = {0, 0, 3, 5, 5}
+ * Values          = {1, 2, 3, 4, 5}
+ * ```
+ * The desired result of this function is the repetition and definition level values that
+ * correspond to the data values:
+ * ```
+ * col = {[], [[], [1, 2, 3], [4, 5]], [[]]}
+ * def = { 0    1,  2, 2, 2,   2, 2,     1 }
+ * rep = { 0,   0,  0, 2, 2,   1, 2,     0 }
+ * ```
+ *
+ * Since repetition and definition levels arrays contain a value for each empty list, the size of
+ * the rep/def level array can be given by
+ * ```
+ * rep_level.size() = size of leaf column + number of empty lists in level 0
+ *                                        + number of empty lists in level 1 ...
+ * ```
+ *
+ * We start with finding the empty lists in the penultimate level and merging it with the indices
+ * of the leaf level. The values for the merge are the definition and repetition levels
+ * ```
+ * empties at level 1 = {0, 5}
+ * def values at 1    = {1, 1}
+ * rep values at 1    = {1, 1}
+ * indices at leaf    = {0, 1, 2, 3, 4}
+ * def values at leaf = {2, 2, 2, 2, 2}
+ * rep values at leaf = {2, 2, 2, 2, 2}
+ * ```
+ *
+ * merged def values  = {1, 2, 2, 2, 2, 2, 1}
+ * merged rep values  = {1, 2, 2, 2, 2, 2, 1}
```

Comment:
```suggestion
 * We start with finding the empty lists in the penultimate level and merging it with the indices
 * of the leaf level. The values for the merge are the definition and repetition levels:
 * ```
 * empty lists at level 1 = {0, 5}
 * definition values at level 1 = {1, 1}
 * repetition values at level 1 = {1, 1}
 * indices at leaf = {0, 1, 2, 3, 4}
 * definition values at leaf = {2, 2, 2, 2, 2}
 * repetition values at leaf = {2, 2, 2, 2, 2}
 *
 * merged def values  = {1, 2, 2, 2, 2, 2, 1}
 * merged rep values  = {1, 2, 2, 2, 2, 2, 1}
 * ```
```